### PR TITLE
2020 04 20 decode message

### PR DESCRIPTION
**Pull request by xkevin190**
work in progress missing order code

### CODE REVIEW

File: components/HTTP/src/Websocket.cpp [hunk 1]
```diff
@@ -47,423 +46,560 @@ static esp_err_t _json_parse_hex(cJSON* root, std::uint8_t *buf,
...
-    char payload[500];
-    esp_err_t err;
-    err = util::decode((std::uint8_t*)buffer, payload);
-    if (err) {
-        ESP_LOGE("TEST", "error decoding the value");
+    for (size_t i = 0; i < len; i++) {
+        const uint8_t upper = (buf[i] & 0xf0) >> 4;
+        const uint8_t lower = (buf[i] & 0x0f);
+        dst[i] = buf[upper];
+        dst[i + 1] = buf[lower];
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.cpp:55]:
```suggestion
   static const char hexmap[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    for (size_t i = 0; i < len; i++) {
        const uint8_t upper = (buf[i] & 0xf0) >> 4;
        const uint8_t lower = (buf[i] & 0x0f);
        dst[i] = hexmap[upper];
        dst[i + 1] = hexmap[lower];
```

hexmap maps `lower` and `upper` to characters

Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:55]:
```suggestion
        dst[i * 2] = buf[upper];
        dst[i * 2 + 1] = buf[lower];
```

File: components/HTTP/src/Websocket.cpp [hunk 2]
```diff
@@ -47,423 +46,560 @@ static esp_err_t _json_parse_hex(cJSON* root, std::uint8_t *buf,
...
+    if (!cbor_value_is_valid(&id_it) &&
+        !cbor_value_is_byte_string(&id_it)) {
+        return ESP_FAIL;
+    }
+
+    /* Verify length is okay */
+    size_t len;
+    cbor_value_calculate_string_length(&id_it, &len);
+    if (len != sizeof(chat_id_t)) {
+        return ESP_FAIL;
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:75]:
```suggestion
        return ESP_ERR_INVALID_SIZE;
```

File: components/HTTP/src/Websocket.cpp [hunk 3]
```diff
@@ -47,423 +46,560 @@ static esp_err_t _json_parse_hex(cJSON* root, std::uint8_t *buf,
...
 
-        err = messageRecipient(ws_pkt.payload, &client_uid);
-        if (err != ESP_OK) {
-            ESP_LOGE(TAG, "Error getting message uids %s", esp_err_to_name(err));
-            return;
-        };
+        /* Parse msgID */
+        if (_parse_chat_id(&it, "msgID", &msg->msg_id) < 0) {
+            ESP_LOGE(TAG, "chat: msgID is invalid!\n");
+            return ESP_FAIL;
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:157]: check if you can use some error code available in: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/error-codes.html

File: components/HTTP/src/Websocket.cpp [hunk 4]
```diff
@@ -47,423 +46,560 @@ static esp_err_t _json_parse_hex(cJSON* root, std::uint8_t *buf,
...
-        };
+        /* Parse message content */
+        if (_parse_msg_content(&it, "msg", &msg->msg) < 0) {
+            ESP_LOGE(TAG, "chat: msg is invalid!\n");
+            return ESP_FAIL;
+        }
+
+        /* Parse timestamp */
+        if (_parse_uint64(&it, "timestamp", &msg->timestamp) < 0) {
+            ESP_LOGE(TAG, "chat: invalid timestamp!\n");
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:168]: check if you can use some error code available in: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/error-codes.html

File: components/HTTP/src/Websocket.cpp (hunk 1 above)
Comment (luisan00) [reply components/HTTP/src/Websocket.cpp:55]:
```suggestion
    for (size_t i = 0; i < len; i++) {
        const uint8_t upper = (buf[i] & 0xf0) >> 4;
        const uint8_t lower = (buf[i] & 0x0f);
        dst[i * 2] = buf[upper];
        dst[i * 2+ 1] = buf[lower];
```

File: components/HTTP/src/Websocket.cpp [hunk 5]
```diff
@@ -464,6 +587,56 @@ esp_err_t Websocket::sendUart(httpd_ws_frame_t ws_pkt)
...
+    esp_err_t err;
+    size_t uidlen = 32;
+    err = parseMessage((std::uint8_t*)buffer, &msg, length);
+    if (err) {
+        ESP_LOGE("TEST", "error decoding the value");
+    }
+
+    cJSON* root = cJSON_CreateObject();
+
+    char* fromUID = (char*)malloc((uidlen * 2) + 1);
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:606]: if you are you allocating memory here you must be free it below

File: components/HTTP/src/Websocket.cpp [hunk 6]
```diff
@@ -464,6 +587,56 @@ esp_err_t Websocket::sendUart(httpd_ws_frame_t ws_pkt)
...
+    ws_pkt.len = strlen(payload);
+    ws_pkt.final = 1;
+
+    // httpd_ws_send_frame(req_handler, &ws_pkt);
+
+    checkMessageType(ws_pkt, true);
+
+
+    // send message to customers
+}
```
Comment (luisan00) [inline components/HTTP/src/Websocket.cpp:642]: add new line

File: components/HTTP/src/Websocket.cpp [hunk 7]
```diff
@@ -56,6 +56,8 @@ esp_err_t bytesToHex(std::uint8_t* buf, char* dst, std::size_t len)
         dst[i * 2 + 1] = hexmap[lower];
     }
 
+    std::cout << "dios meo" << dst << std::endl;
```
Comment (jeandudey) [inline components/HTTP/src/Websocket.cpp]:
```suggestion
    ESP_LOGW(TAG, "why this doens't work: %s!!!", dst);
```

Comment (xkevin190) [reply components/HTTP/src/Websocket.cpp]: this was resolved in the commit 2dd4a1f9fa14aebe92dea0cfc885e000929295c0

**Pull request merged**